src/meta/MetaHttpDownloadHandler.cpp

```diff
@@ -0,0 +1,237 @@
+    std::vector<std::string> files;
+    folly::split("\n", result.value(), files, true);
+    int32_t  partNumber = files.size() - 1;
+
+    std::vector<cpp2::HostItem> hostItems;
+    std::unique_ptr<kvstore::KVIterator> hostIter;
+    auto hostPrefix = MetaServiceUtils::hostPrefix();
+    auto hostRet = kvstore_->prefix(0, 0, hostPrefix, &hostIter);
+    if (hostRet != kvstore::ResultCode::SUCCEEDED) {
+        return false;
+    }
+
+    std::unordered_map<HostAddr, std::vector<PartitionID>> hostPartition;
+    while (hostIter->valid()) {
+        cpp2::HostItem item;
+        nebula::cpp2::HostAddr host;
+        auto hostAddrPiece = hostIter->key().subpiece(hostPrefix.size());
+        memcpy(&host, hostAddrPiece.data(), hostAddrPiece.size());
+        item.set_hostAddr(host);
+        auto address = std::make_pair(host.get_ip(), host.get_port());
+        auto hostIter_ = hostPartition.find(address);
+        if (hostIter_ == hostPartition.end()) {
+            std::vector<PartitionID> partitions;
+            hostPartition.insert(std::make_pair(address, partitions));
+        }
+        hostItems.emplace_back(item);
+        hostIter->next();
+    }
+
+    std::unique_ptr<kvstore::KVIterator> iter;
+    auto prefix = MetaServiceUtils::partPrefix(spaceID_);
+    auto ret = kvstore_->prefix(0, 0, prefix, &iter);
+    if (ret != kvstore::ResultCode::SUCCEEDED) {
+        return false;
+    }
+
+    int32_t partSize{0};
+    while (iter->valid()) {
+        auto key = iter->key();
+        PartitionID partId;
+        memcpy(&partId, key.data() + prefix.size(), sizeof(PartitionID));
+        for (auto host : MetaServiceUtils::parsePartVal(iter->val())) {
+            auto address = std::make_pair(host.get_ip(), host.get_port());
+            hostPartition[address].emplace_back(partId);
+        }
+        partSize++;
+        iter->next();
+    }
+
+    if (partNumber != partSize) {
+        LOG(ERROR) << "HDFS part number should be equal with nebula "
+                   << partNumber << " " << partSize;
+        return false;
+    }
+
+    std::atomic<int> completed(0);
+    std::vector<std::thread> threads;
+    for (auto &pair : hostPartition) {
+        std::vector<PartitionID> partitions;
+        for (auto part : pair.second) {
+            partitions.emplace_back(part - 1);
+        }
+        std::string partNumbers;
+        folly::join(",", partitions, partNumbers);
+
+        auto host = network::NetworkUtils::intToIPv4(pair.first.first);
+        auto t = "http://%s:%d/download?url=%s&port=%d&path=%s&parts=%s&local=%s";
+        auto download = folly::stringPrintf(t, host.c_str(), FLAGS_storage_http_port, url.c_str(),
+                                            port, path.c_str(), partNumbers.c_str(), local.c_str());
+        command = folly::stringPrintf("/usr/bin/curl -G \"%s\" 2> /dev/null", download.c_str());
+        LOG(INFO) << "Fetch Storage: " << download;
+        threads.push_back(std::thread([&]() {
```

Comment:
why not limit the thread number?

Reply:
I wish to process download in storage services at the same time

Reply:
because I do not know how many host in hostPartition at most, if there are 100 hosts, then start so many thread may take up too many cpu and network resource and will stop other service in this server.

Reply:
`MetaService` just send a HTTP request to notify `StorageService` start download . It's seems won't cast lots of resource.

If I using thread pool and some storage have finished and other not start. It will very strange.